streaming-binlog/src/main/java/org/openmrs/analytics/DebeziumListener.java

```diff
@@ -0,0 +1,132 @@
+                        constant(Envelope.Operation.UPDATE.code()));
+
+        final Predicate isCreateEvent =
+                header(DebeziumConstants.HEADER_OPERATION).in(
+                        constant(Envelope.Operation.READ.code()),
+                        constant(Envelope.Operation.CREATE.code()));
+
+        final Predicate isUpdateEvent =
+                header(DebeziumConstants.HEADER_OPERATION).in(
+                        constant(Envelope.Operation.READ.code()),
+                        constant(Envelope.Operation.UPDATE.code()));
+
+        final Predicate isDeleteEvent =
+                header(DebeziumConstants.HEADER_OPERATION).in(
+                        constant(Envelope.Operation.READ.code()),
+                        constant(Envelope.Operation.DELETE.code()));
+
+        /**
+         * Main Change Data Capture (DBZ) entrypoint
+         * @param dbzConfig.
+         */
+        from(getDebeziumConfig())
+                .routeId(DebeziumListener.class.getName() + ".MysqlDatabaseCDC")
+                .log(LoggingLevel.INFO, "Incoming Events: ${body} with headers ${headers}")
+                .choice()
+                    .when(isCreateEvent)
+                        .log(LoggingLevel.TRACE, "CreateEvent Emitted ---> ${body}")
+                        .process(new FhirUriGenerator())
+                        .to(FHIR_ROUTE_HANDLER)
+                        .endChoice()
+                    .when(isUpdateEvent)
+                        .log(LoggingLevel.TRACE, "UpdateEvent Emitted ---> ${body}")
+                        .process(new FhirUriGenerator())
+                        .to(FHIR_ROUTE_HANDLER)
+                        .endChoice()
+                    .when(isDeleteEvent)
+                        .log(LoggingLevel.TRACE, "DeleteEvent Emitted Not Supported ---> ${body}")
+                        .endChoice()
+                    .otherwise()
+                        .log(LoggingLevel.WARN, "Event Not Supported: ${headers[" + DebeziumConstants.HEADER_IDENTIFIER + "]}")
+                .endParent();
+
+        /**
+         * FHIR event worker
+         * Current append mode is suppported i.e any Create or Update will be upserted to the warehouse
+         * append mode should also work for deletion given that openmrs does not delete data instead voids
+         * @param routeDefinition.
+         */
+        from(FHIR_ROUTE_HANDLER)
+                .filter(body().isNotNull())// Filter non-Fhir uri
+                .choice()
+                    .when().simple("${header.fhirResourceUri} == null || ${header.fhirResourceUri} == ''")
+                        .log(LoggingLevel.WARN,"FHIR URL Not Mapped --->  ${headers[" + DebeziumConstants.HEADER_IDENTIFIER + "]}")
+                    .otherwise()
+                        .setBody(simple("${null}")) // Set up body to null
+                        .toD("{{openmrs.serverUrl}}{{openmrs.fhirBaseEndpoint}}${header.fhirResourceUri}" +
```

Comment:
Following up on my comment above about Camel and code readability, I particularly feel uneasy about these strings that have code templates in them (e.g., here, or `simple` call above, etc.). Is it possible to avoid these and move this kind of logic to the `process` method as much as possible? I mean can we make the decision to drop a message and/or send a request to OpenMRS FHIR endpoint with direct Java code inside `process` below?

Reply:
Very true! I have removed all template (DSL) expression and used pure java code! Moving forward I think we should ensure we avoid camel DSL expressions as much as possible!